net: openthread: Allow to configure Sleepy End Device

rlubos opened the pull request:
<s>This PR depends on #15521, hence DNM until it's merged.</s>

Currently, OpenThread does not support Sleepy End Devices correctly, due to two reasons:

1. OpenThread radio layer in Zephyr does not process the actual ACK frame received on 15.4 radio but provides a fake one to OpenThread. Because of that, all ACK frames provided to OpenThread have `Frame Pending` field set to 0, breaking the Sleepy End Device operation - it will never wait for more data to come, and will always put the radio to sleep instead.

2. OpenThread radio layer does not implement `Frame Pending` bit management API required by OpenThread. Because of that, Zephyr routers do not set this field correctly in ACK response, even when they have more data to send to a sleeping child. The outcome is the same as in 1. - sleeping child will put the radio to sleep instead of waiting for more data.

This PR fixes the former problem. The latter one, somehow covered by #12942, is more complex and will require to update `ieee802154_radio` API with some ACK management functions, as radio drivers tend to send ACK responses automatically. I should be able to propose something soon.

rlubos committed:
net: openthread: Allow to configure SED

Allow to select Sleepy End Device, and configure it during OpenThread
initialization.

According to Thread Specification, Sleepy End Devices should always
attach to the network as SED, to indicate increased buffer requirement
to a parent. Therefore, we reconfigure the Link Mode on each boot.

Note, that Poll Period value is not stored in the persistent storage,
hence we also need to initialize it on each boot.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
drivers: ieee802154_nrf5: Add ACK handling

nRF5 driver did not utilize `ieee802154_radio_handle_ack` API, therefore
did not provide ACK frames to the upper layer. This commit fixes this
problem.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: openthread: Process real ACK frame instead of fake one

OpenThread radio layer did not implement `ieee802154_radio_handle_ack`
API and provided fake ACK frame to the OpenThread.

This prevented proper Sleepy End Device operation, as it expects to
receive information in the ACK whether it should wait for more data to
come or should it put the radio to sleep.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
Dependencies got merged, rebased and removed DNM.

jukkar reviewed (approved):
LGTM

mike-scott reviewed (approved):
Haven't tested, but approach looks good.

mike-scott commented:
@rlubos What about other 802.15.4 drivers who don't implement ACK handling?  Does this affect them negatively?

mike-scott commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -253,6 +253,46 @@ static int nrf5_set_txpower(struct device *dev, s16_t dbm)
+	net_pkt_cursor_init(ack_pkt);
+
+	if (ieee802154_radio_handle_ack(nrf5_radio->iface, ack_pkt) != NET_OK) {
+		LOG_INF("ACK pakcet not handled - releasing.");
```
Typo here.

rlubos replied:
Thanks, fixed.

rlubos commented:
@mike-scott The requirement now is to call `ieee802154_radio_handle_ack` API provided by Zephyr to pass the ACK frame to the upper layer. Otherwise, OpenThread will consider every transmission as failed (no ACK received). So yes, this will impact negatively drivers that do not use this API. 

From what I've seen, the only driver that does not implement this (beside nRF, which is fixed in this PR) is KW41Z driver. I'm not sure who is responsible for this driver. I saw that @dleach02 had some contributions regarding OpenThread, I'll add him to reviewers.

rlubos commented:
@mike-scott Any other feedback on this?

mike-scott commented:
@rlubos Actually I was hoping @dleach02 or @MaureenHelm would weigh in about KW41Z taking this approach as well.   The only objection I would have to this is if it breaks another driver w/o giving them time to update.

jukkar commented:
@mike-scott @rlubos Any chance this could get to 2.0?

rlubos commented:
@jukkar It's not really a question to me, this PR is waiting for NXP feedback on ACK support in KW41Z  driver. From my point of view, it's good to go, just another small step towards making OpenThread certifiable with Zephyr.

jukkar commented:
> It's not really a question to me, this PR is waiting for NXP feedback on ACK support in KW41Z driver. From my point of view, it's good to go, just another small step towards making OpenThread certifiable with Zephyr.

IMHO this looks good to go too. @mike-scott can you remove your -1?

pfalcon commented on drivers/ieee802154/ieee802154_nrf5.c:
```diff
@@ -253,6 +253,46 @@ static int nrf5_set_txpower(struct device *dev, s16_t dbm)
+		goto free_nrf_ack;
+	}
+
+	if (net_pkt_write(ack_pkt, nrf5_radio->ack_frame.psdu + 1,
```
For an outsider, it's of course not clear why you offset "raw" buffer by 1 here.

rlubos replied:
Will add comment on that.

retfie replied:
@mike-scott isn't your request for change (typo) wasn't fixed ? Could you remove your -1 ?

mike-scott replied:
Thank you for the ping.  Approved.

pfalcon commented on subsys/net/l2/openthread/Kconfig:
```diff
@@ -129,6 +129,15 @@ config OPENTHREAD_MTD
+
+config OPENTHREAD_POLL_PERIOD
+	int "Poll period for sleepy end devices [ms]"
+	default 236000
```
236 seconds? Is it good default? (Might be good value for *some* production devices (hmm, I personally wouldn't like to own a temperature sensor which takes 4mins to update), but unlikely a good default.

rlubos replied:
Thread specification defaults the child timeout to 240 seconds. A sleepy end device needs to poll parent within this period, otherwise, it will be dropped. 4 extra seconds are for possible retry attempts. It's the same default as we use in OpenThread.

And just to clarify - the temperature sensor scenario you brought up is not a good example of how sleepy end devices poll mechanism work. A sleepy end device polls the parent (powered router) periodically, to obtain any data the parent queued for it. There are no restrictions on when a sleepy child can send data to its parent - it can do it as often as it wants (e.g. when temperature value changes above some configured threshold). Of course, if the sleepy end device expects a response for its notification (for instance ACK on CoAP), it should either decrease the poll period or switch to MTD mode (non-sleepy) until response is recevied.

pfalcon replied:
@rlubos, Thanks for explanation. So, it's effectively more like a timeout than a period, though OpenThread terminology prefers the latter.

pfalcon commented on subsys/net/lib/openthread/platform/radio.c:
```diff
@@ -50,6 +56,38 @@ static struct ieee802154_radio_api *radio_api;
+	size_t ack_len = net_pkt_get_len(pkt);
+
+	if (ack_len != ACK_PKT_LENGTH) {
+		return NET_CONTINUE;
```
Log something?

rlubos replied:
It could lead to a log flood on some devices. While it could work with nRF driver, which calls this function explicitly for ACK frames, there are drivers that call `ieee802154_radio_handle_ack` for every single frame, and decide whether it was an ACK frame based on the result. For instance [here](https://github.com/zephyrproject-rtos/zephyr/blob/f2cb20c772b3700896cf0f67b7e307677866058e/drivers/ieee802154/ieee802154_uart_pipe.c) (and it's not a sole example).

pfalcon commented on subsys/net/lib/openthread/platform/radio.c:
```diff
@@ -50,6 +56,38 @@ static struct ieee802154_radio_api *radio_api;
+	}
+
+	if ((*net_pkt_data(pkt) & FRAME_TYPE_MASK) != FRAME_TYPE_ACK) {
+		return NET_CONTINUE;
```
Log something?

pfalcon commented on subsys/net/lib/openthread/platform/radio.c:
```diff
@@ -36,10 +36,16 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 static otRadioState sState = OT_RADIO_STATE_DISABLED;
 
 static otRadioFrame sTransmitFrame;
+static otRadioFrame ack_frame;
```
How is this static buffer going to work?

rlubos replied:
It just holds a copy of the ACK frame until it's processed by OpenThread.

pfalcon replied:
> It just holds a copy of the ACK frame until it's processed by OpenThread.

So, is this processing is 100% cooperative multitasking driven, even in the case of multiple OT radios? That's the only scenario when a static buffer would work, and I'd be surprised to learn that such processing is used on top of otherwise preemptively multitasked Zephyr (and such support would require a lot of checks).

rlubos replied:
Only a single OT radio instance is supported. If ACK was requested, the radio tx call will block until the ACK is received or a timeout occurs. And I assumed here that radio driver is smart enough not to notify another ACK until another transmit is called (that would need some extra care). Do we need anything else here?

pfalcon replied:
> Do we need anything else here?

No, just wanted to be sure that these points were raised in the review and some responses given. Let me approve from my side.

dleach02 commented:
I'm confused about how building an artificial ack frame to pass up the stack handles the problem of passing the "Frame Pending" status in an ACK received on a TX? 

When I look at this code it isn't clear to me where it may be doing something with this status bit?

rlubos commented:
> I'm confused about how building an artificial ack frame to pass up the stack handles the problem of passing the "Frame Pending" status in an ACK received on a TX?

The whole point of this PR is to provide the actual ACK frame received on the radio instead of an artificial one, we've provided so far.

> When I look at this code it isn't clear to me where it may be doing something with this status bit?

Because we don't do anything with the bits received in the ACK. It's just copied and passed to the OpenThread stack, which makes use of it.

> Does nrf5_radio->ack_frame.psdu point to the actual ACK frame received on the Nordic parts?

Yes, the `nrf_802154_transmitted_raw` callback provides the ack frame (if received) and `nrf5_radio->ack_frame.psdu` is set to point to it.

dleach02 commented:
okay. The last part was what I missed. Clearly if you are basically just passing up the ack frame then it will contain the frame pending bit information.

dleach02 commented:
The KW41Z can report up the state of the FP bit. Since the implementation here wants an ACK frame sent up the driver will need to construct an ACK frame to report up the FP bit as the ACK was basically consumed by the HW with the status of the bit saved in a register.

rlubos committed:
samples: net: Remove redundant frdm_kw41z configs from echo samples

echo_client and echo_server configuration for 802.15.4 and OpenThread is
covered by overlay files and default board configuration. Board-specific
configuration for frdm_kw41z was redundant in this case.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
drivers: ieee802154_kw41z: Create a fake ACK frame for upper layers

Upper layers like OpenThread expect the radio driver to forward ACK
frame received by the radio.
Simulate this behavior on kw41z by recreating the ACK frame from the
available data.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos commented:
I've added ACK handling for the kw41z driver. In case ACK was received, the driver will re-create the ACK frame from the data available and forward it to the upper layer. Also, it turned out that the OpenThread variant of the samples did not build for the frdm_kw41z, due to conflicting board config file. As these board-specific configurations duplicated with overlay-802154.conf, I've removed them.

Note that I have no hardware to verify samples with this board. All I can assure is that the samples build fine for it.

@dleach02 @mike-scott Please revisit this PR.

mike-scott reviewed (approved):
LGTM
